components/builder-core/src/api.rs

```diff
@@ -0,0 +1,275 @@
+    let mut channel_req = OriginChannelGet::new();
+    channel_req.set_origin_name(ident.get_origin().to_string());
+    channel_req.set_name(channel.to_string());
+
+    match conn.route::<OriginChannelGet, OriginChannel>(&channel_req) {
+        Ok(origin_channel) => {
+            if !check_origin_access(session_id, ident.get_origin())
+                .map_err(Error::NetError)?
+            {
+                return Err(Error::OriginAccessDenied);
+            }
+
+            let mut request = OriginPackageGet::new();
+            request.set_ident(ident.clone());
+            match conn.route::<OriginPackageGet, OriginPackage>(&request) {
+                Ok(package) => {
+                    let mut promote = OriginPackagePromote::new();
+                    promote.set_channel_id(origin_channel.get_id());
+                    promote.set_package_id(package.get_id());
+                    promote.set_ident(ident.clone());
+                    match conn.route::<OriginPackagePromote, NetOk>(&promote) {
+                        Ok(_) => Ok(()),
+                        Err(err) => Err(Error::NetError(err)),
+                    }
+                }
+                Err(err) => Err(Error::NetError(err)),
+            }
+        }
+        Err(err) => Err(Error::NetError(err)),
+    }
+}
+
+pub fn promote_job_group_to_channel(group_id: u64, channel: &str, session_id: u64) -> Result<()> {
+    let mut group_get = GroupGet::new();
+    group_get.set_group_id(group_id);
+
+    let mut conn = Broker::connect().unwrap();
+    let group = match conn.route::<GroupGet, Group>(&group_get) {
+        Ok(g) => g,
+        Err(err) => return Err(Error::NetError(err)),
+    };
+
+    // This only makes sense if the group is complete. If the group isn't complete, return now and
+    // let the user know.
+    if group.get_state() != GroupState::Complete {
+        return Err(Error::GroupNotComplete);
+    }
+
+    let mut failed_projects = Vec::new();
+    let mut origin_map = HashMap::new();
+
+    // We can't assume that every project in the group belongs to the same origin. It's entirely
+    // possible that there are multiple origins present within the group. Because of this, there's
+    // no way to atomically commit the entire promotion at once. It's possible origin shards can be
+    // on different machines, so for now, the best we can do is partition the projects by origin,
+    // and commit each origin at once. Ultimately, it'd be nice to have a way to atomically commit
+    // the entire promotion at once, but that would require a cross-shard tool that we don't
+    // currently have.
+    for project in group.get_projects().into_iter() {
+        if project.get_state() == ProjectState::Success {
+            let ident = OriginPackageIdent::from_str(project.get_ident()).unwrap();
+
+            if channel != "stable" {
+                match create_channel(ident.get_origin(), channel, session_id) {
```

Comment:
We can't move it from where it's at because the origin to create the channel for comes from the current package in the iteration, and there's no way to guarantee that every package will have the same origin.